base/math.jl

```diff
@@ -536,38 +536,52 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+	#= Keep specialized two argument version, because it is faster
+	than generic varargs version below =#
+	ay, ax = minmax(abs(x), abs(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
     rr = ax * sqrt(1 + r * r)
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
-    if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
-    else
-        return rr
-    end
+	if isnan(r)
+		if isinf(x) || isinf(y)
+			return oftype(rr, Inf)
+		else
+			return oftype(rr, NaN)
+		end
+   	else
+   		return rr
+   	end
 end
-
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = _hypot(promote(x...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
+	N == 0 && error("invalid hypot() call without arguments")
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	r = sqrt(s)
```

Comment:
Mustn't the type of `r = sqrt(s)` exactly match that of `s`, given that `s` is always a positive float (since `s = sum(abs2, float.(x))`). (similarly, if `s` is `Inf` or `NaN`, then `r = sqrt(s)` will also be)

Reply:
Ah your point is that `typeof(r) == typeof(s)`?